lib/ansible/modules/network/panos/panos_dag_tags.py

```diff
@@ -0,0 +1,253 @@
+    HAS_LIB = False
+
+
+def get_devicegroup(device, devicegroup):
+    dg_list = device.refresh_devices()
+    for group in dg_list:
+        if isinstance(group, panorama.DeviceGroup):
+            if group.name == devicegroup:
+                return group
+    return False
+
+
+def register_ip_to_tag_map(device, ip_addresses, tag):
+    """
+    :param device:
+    :param ip_addresses:
+    :param tag:
+    :return:
+    """
+
+    exc = None
+    try:
+        device.userid.register(ip_addresses, tag)
+    except Exception:
+        exc = get_exception()
+        return False, exc
+
+    if exc:
+        return False, exc
+    else:
+        return True, exc
+
+
+def get_all_address_group_mapping(device):
+    """
+    Retrieve all the tag to IP address mappings
+    :param device:
+    :return:
+    """
+
+    exc = None
+    ret = None
+    try:
+        ret = device.userid.get_registered_ip()
+    except Exception:
+        exc = get_exception()
+
+    if exc:
+        return False, exc
+    else:
+        return ret, exc
+
+
+def delete_address_from_mapping(device, ip_address, tags):
+    """
+    Delete an IP address from a tag mapping.
+    :param device:
+    :param ip_address:
+    :param tags:
+    :return:
+    """
+
+    exc = None
+    try:
+        ret = device.userid.unregister(ip_address, tags)
+    except Exception:
+        exc = get_exception()
+
+    if exc:
+        return False, exc
+    else:
+        return True, exc
+
+
+def main():
+    argument_spec = dict(
+        ip_address=dict(required=True),
+        password=dict(required=True),
+        username=dict(default='admin'),
+        api_key=dict(no_log=True),
+        devicegroup=dict(default=None),
+        description=dict(default=None),
+        ip_to_register=dict(type='str', required=False),
+        tag_names=dict(type='list', required=True),
+        commit=dict(type='bool', default=True),
+        operation=dict(type='str', required=True)
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)
+    if not HAS_LIB:
+        module.fail_json(msg='pan-python is required for this module')
+
+    ip_address = module.params["ip_address"]
+    password = module.params["password"]
+    username = module.params['username']
+    api_key = module.params['api_key']
+    commit = module.params['commit']
+    devicegroup = module.params['devicegroup']
+    operation = module.params['operation']
+
+    # Create the device with the appropriate pandevice type
+    device = base.PanDevice.create_from_device(ip_address, username, password, api_key=api_key)
+
+    # If Panorama, validate the devicegroup
+    dev_group = None
+    if devicegroup and isinstance(device, panorama.Panorama):
+        dev_group = get_devicegroup(device, devicegroup)
+        if dev_group:
+            device.add(dev_group)
+        else:
+            module.fail_json(msg='\'%s\' device group not found in Panorama. Is the name correct?' % devicegroup)
```

Comment:
`msg="'%s' device group not found in Panorama. Is the name correct?" % devicegroup)`

Reply:
ok